src/nvim/CMakeLists.txt

```diff
@@ -252,6 +268,17 @@ add_custom_command(OUTPUT ${GENERATED_OPTIONS}
   DEPENDS ${OPTIONS_GENERATOR} ${OPTIONS_LIST_FILE}
 )
 
+# Make NEOVIM_GENERATED_FOR_SOURCES and NEOVIM_GENERATED_FOR_HEADERS mutually exclusive
+# If a header in NEOVIM_GENERATED_FOR_HEADERS is also included in NEOVIM_GENERATED_SOURCES,
+# the header should be included only in NEOVIM_GENERATED_HEADERS
+
+foreach(hfile ${NEOVIM_GENERATED_FOR_HEADERS})
+  list(FIND NEOVIM_GENERATED_FOR_SOURCES ${hfile} hfile_idx)
+  if(NOT ${hfile_idx} EQUAL -1)
+    list(REMOVE_AT NEOVIM_GENERATED_FOR_SOURCES ${hfile_idx})
```

Comment:
I think the idea is easier understandable with example:

```C
/// @file foo.generated.h
const char foo_module[] = { 42, 42, 0 };
static const char foo_module_static[] = { 42, 42, 0 };
//#define FOO_MACROS "bar"
typedef struct {int a;} FooType;
static inline int foo_func(void) { return 0; }
int foo_func_2(void) { return 0; }
```

```C
/// @file foo.h
#ifndef FOO_H
#define FOO_H
#include "foo.generated.h"
#endif  // FOO_H
```

```C
/// @file foo.c
#include "foo.h"
#include "foo.generated.h"
int main(const int argc, const char *const *const argv, const char *const *const env)
{
  return 0;
}
```

Try making this compile. Blasting any line from `foo.generated.h` except any single non-comment non-blank line is permitted, but there must be at least one of this. The single thing out of that file that works is macro redefinition due to some [weird rules](https://gcc.gnu.org/onlinedocs/cpp/Undefining-and-Redefining-Macros.html) which allow redefining macros with the same macros without a warning (guess there is some good historical reason for this) (C99 standard also allows this, this is not something gcc authors designed on themselves). Which is why I commented that macro definition out (in any case we don’t normally define macros in generated files).